Add documentation for moderated sessions

xacrimon opened the pull request:
This PR adds preview documentation for #8563.

xacrimon committed:
add reference bit

xacrimon committed:
getting started

xacrimon committed:
improve tdocs

ptgott commented on docs/pages/preview/moderated-sessions.mdx:
```diff
@@ -0,0 +1,192 @@
+
+Previously, Server Access did not have access control for whom can join a session.
+To work around this, RBAC are ignored for users that only have V4 roles.
+New and modified roles are created as V5 roles any having any attached V5 role will
```
I had trouble following this sentence. Could you please write it in a different way?

ptgott commented on docs/pages/preview/moderated-sessions.mdx:
```diff
@@ -0,0 +1,192 @@
+to start it. The policy applies to SSH and Kubernetes sessions only.
+
+When a user with this require policy starts a session, it will be pending
+until the policy is fulfilled, only then does the session start.
```
I think we can remove ", only then does the session start"

ptgott commented on docs/pages/preview/moderated-sessions.mdx:
```diff
@@ -0,0 +1,192 @@
+spec:
+  allow:
+    require_session_join:
+      - name: Audtor oversight
```
"Audtor" -> "Auditor"

ptgott commented on docs/pages/preview/moderated-sessions.mdx:
```diff
@@ -0,0 +1,192 @@
+
+## MFA-based presence
+
+When `mfa_required` is set to `true` via role or cluster settings. Teleport enforces
```
". Teleport enforces" -> ", Teleport enforces"

ptgott commented on docs/pages/preview/moderated-sessions.mdx:
```diff
@@ -0,0 +1,192 @@
+When `mfa_required` is set to `true` via role or cluster settings. Teleport enforces
+MFA-based presence checks for moderators.
+
+This requires that all moderators wishing to join have a configured U2F or Webauthn MFA token.
```
- Webauthn -> WebAuthn
- Let's combine this with the previous paragraph

ptgott commented on docs/pages/preview/moderated-sessions.mdx:
```diff
@@ -0,0 +1,192 @@
+MFA-based presence checks for moderators.
+
+This requires that all moderators wishing to join have a configured U2F or Webauthn MFA token.
+Every 30 seconds, Teleport will issue a prompt to the user in the terminal, asking them
```
This should be a new paragraph

ptgott commented on docs/pages/preview/moderated-sessions.mdx:
```diff
@@ -0,0 +1,192 @@
+This requires that all moderators wishing to join have a configured U2F or Webauthn MFA token.
+Every 30 seconds, Teleport will issue a prompt to the user in the terminal, asking them
+to press their MFA token in the next 15 seconds. This will happen continously during the session
+and exists so that moderators are always present and watching a given session. If no MFA input
```
New paragraph at "If no MFA input"

xacrimon replied:
Resolved

ptgott reviewed (commented):
Just confirming whether the intention is to make this page unreachable from the main navigation (i.e., the URL will be shared directly with prospects/customers)

xacrimon commented on docs/pages/preview/moderated-sessions.mdx:
````diff
@@ -0,0 +1,192 @@
+tsh kube join $SESSION_ID
+```
+
+## RBAC
````
Yep, at least looking at old preview pages. The preview docs are usually shorter and demonstrate how to get started and maybe try something unreleased out with the intent to give a somewhat accurate description of how something works, what constraints it has and the workflow for using it.

ptgott replied:
👍

xacrimon committed:
address preview docs feedback

xacrimon commented on docs/pages/preview/moderated-sessions.mdx:
```diff
@@ -0,0 +1,192 @@
+### Filters
+
+Require policies can specify which users they consider as valid with a filter expression. They work like
+other filters used within Teleport such with access requests. The filter context has a `user` object defined
```
I don't think we have an all purpose guide to them. Would it be better to give a short example and then link another page which has more detailed information?

ptgott replied:
An example with a short explanation sounds good. I don't think we have a good page we can link to here, so I'll create an issue to create a guide to filter syntax.

xacrimon committed:
fix nav

xacrimon commented:
@ptgott I've added the page to the navbar also.

ptgott commented on docs/pages/preview/moderated-sessions.mdx:
```diff
@@ -0,0 +1,200 @@
+The system is based around **require policies** and allow policies. Require policies define a set of conditions
+that must be a met for a session to start/run.
+
+**Allow policies** are used to define what a user can do. In essence, what sessions they can join
```
```suggestion
**Allow policies** are used to define what sessions a user can join
```

ptgott commented on docs/pages/preview/moderated-sessions.mdx:
```diff
@@ -0,0 +1,200 @@
+that must be a met for a session to start/run.
+
+**Allow policies** are used to define what a user can do. In essence, what sessions they can join
+and under what conditions they may join the session.
```
```suggestion
and under what conditions they may join a session.
```

ptgott commented on docs/pages/preview/moderated-sessions.mdx:
```diff
@@ -0,0 +1,200 @@
+### Backwards compatibility with Server Access
+
+Previously, Server Access did not include controls over which users can join a session.
+To work around this, RBAC rules are ignored for users that only have V4 roles.
```
```suggestion
To work around this, RBAC rules are ignored for users that only have V4 roles (`version: v4` in the role specification).
```

ptgott commented on docs/pages/preview/moderated-sessions.mdx:
```diff
@@ -0,0 +1,200 @@
+
+Previously, Server Access did not include controls over which users can join a session.
+To work around this, RBAC rules are ignored for users that only have V4 roles.
+New roles are created as V5 and V4 roles are also upgraded when they are modified in the UI.
```
```suggestion
New roles are created as V5. V4 roles are upgraded when they are modified in the UI.
```
I originally mis-parsed this as "as (V5 and V4 roles)," so I've suggested this to make this clearer.

klizhentas commented on docs/pages/preview/moderated-sessions.mdx:
```diff
@@ -0,0 +1,200 @@
+Moderated sessions allows Teleport administrators to
+define requirements for other users to be present in a Server or Kubernetes Access session.
+
+Moderated Sessions is useful in the following scenarios:
```
Nit: I'd say Moderated Sessions Feature is useful or Moderated sessions are useful

klizhentas commented on docs/pages/preview/moderated-sessions.mdx:
```diff
@@ -0,0 +1,200 @@
+h1: Moderated Sessions
+---
+
+This is an upcoming feature expected to land in Teleport 8.2.
```
Please use `Details` with min version:

```
<Details title="Details title" min="8.2" opened>
  This is an upcoming feature expected to land in Teleport 8.2.
</Details>
```

klizhentas commented on docs/pages/preview/moderated-sessions.mdx:
```diff
@@ -0,0 +1,200 @@
+spec:
+  allow:
+    kubernetes_groups:
+      - system:masters
```
that's anti-pattern, we should not recommend folks use this admin role by default, let's pick another role, like view

klizhentas commented on docs/pages/preview/moderated-sessions.mdx:
````diff
@@ -0,0 +1,200 @@
+
+Please substitute `$SESSION_ID` with the UUID shown when starting the session.
+
+```sh
````
use `code` block here and below

klizhentas commented on docs/pages/preview/moderated-sessions.mdx:
````diff
@@ -0,0 +1,200 @@
+```sh
+# adam
+
+tsh kube login minikube
````
add $ sign in code block here and below

klizhentas commented on docs/pages/preview/moderated-sessions.mdx:
````diff
@@ -0,0 +1,200 @@
+tsh kube join $SESSION_ID
+```
+
+## RBAC
````
Let's move this RBAC section into Setup -> Reference -> Moderated Sessions.

The steps above are a simple step by step guide with no details, the reference dives into all the details and options. 

cc @ptgott @xinding33 @xacrimon

xacrimon replied:
Since moderated sessions isn't far out I've decided to convert this from a preview doc to a regular doc. I've created a reference file in addition to a quick setup guide under access controls.

klizhentas reviewed (changes requested):
Overall looks good, but let's split the guide and reference

xacrimon committed:
Apply suggestions from code review

Co-authored-by: Paul Gottschling <paul.gottschling@goteleport.com>

xacrimon committed:
content specific feedback

xacrimon committed:
rework docs

xacrimon committed:
add semicolon

xacrimon committed:
restore submodules

ptgott commented on docs/pages/access-controls/guides/moderated-sessions.mdx:
````diff
@@ -0,0 +1,99 @@
+
+### Step 1/4: Create the prod-access role
+
+```code
````
This snippet is going to look a bit odd until [this issue](https://github.com/gravitational/next/issues/811) is fixed, but I don't think we should block on that issue.

xacrimon replied:
Agreed, not much we can do here except continue

ptgott commented on docs/pages/setup/reference/moderated-sessions.mdx:
```diff
@@ -0,0 +1,115 @@
+Moderated Sessions makes use of RBAC policies to allow for fine grained control over
+who can join a session and who are required to be present to start one.
+
+The system is based around **require policies** and allow policies. Require policies define a set of conditions
```
```suggestion
The system is based around **require policies** and **allow policies**. 

Require policies define a set of conditions
```

Since we introduce both concepts in the same sentence, I think it looks a bit cleaner to separate it from the sentence that goes in depth on require policies.

ptgott commented on docs/pages/setup/reference/moderated-sessions.mdx:
```diff
@@ -0,0 +1,115 @@
+The system is based around **require policies** and allow policies. Require policies define a set of conditions
+that must be a met for a session to start/run.
+
+**Allow policies** are used to define what sessions a user can join
```
```suggestion
Allow policies are used to define what sessions a user can join
```
Suggested removing the bold here and bolding the term when we first introduced it.

ptgott commented on docs/pages/setup/reference/moderated-sessions.mdx:
```diff
@@ -0,0 +1,115 @@
+
+### Session kinds
+
+Require and allow policies have to specify which sessions they apply to. Valid options are `ssh` and `k8s`.
```
Which config field would you use these values for?

xacrimon replied:
I've moved the section down and pointed to the example configs on how to specify the kinds.

ptgott commented on docs/pages/setup/reference/moderated-sessions.mdx:
```diff
@@ -0,0 +1,115 @@
+
+Require policies can specify which users they consider as valid with a filter expression. They work like
+other filters used within Teleport such with access requests. The filter context has a `user` object defined
+which the set fields `roles`, `traits` and `name`.
```
While I know a full example is coming up in this page, I think other readers may be expecting an example here. Otherwise, it may not be clear where a filter appears within a config resource.

xacrimon replied:
Do you mean just a filter example or like a minimal policy with a filter? The reason I am hesitant is that it'll be pretty much the exact same as the examples below since there aren't really optional fields on a policy.

ptgott commented on docs/pages/setup/reference/moderated-sessions.mdx:
```diff
@@ -0,0 +1,115 @@
+
+**Allow policies** are used to define what sessions a user can join
+and under what conditions they may join a session.
+
```
Since I've suggested using the current "RBAC" section intro as the intro for the whole page, let's add a new H2 heading here. Maybe "Configuring Moderated Sessions"? This will include the H3 headings re: specific config options.

```suggestion

## Configuring Moderated Sessions
```

ptgott commented on docs/pages/setup/reference/moderated-sessions.mdx:
```diff
@@ -0,0 +1,115 @@
+
+## MFA-based presence
+
+When `mfa_required` is set to `true` via role or cluster settings, Teleport enforces
```
- If MFA-based presence is another RBAC config setting, should we include it in the previous section (i.e., as an H3 heading) rather than as a new H2-level section?
- I think it would be helpful to add examples of what `mfa_required` looks like in role or cluster settings. This would make it clear where in a config spec we would expect to find this field (e.g., directly under `spec` or under another field).

xacrimon replied:
I got the MFA setting name wrong. We've got a page on per session MFA which is now linked.

ptgott commented on docs/pages/setup/reference/moderated-sessions.mdx:
```diff
@@ -0,0 +1,115 @@
+### Participant Modes
+
+A participant joining a session will always have one of three modes:
+- Peers can join and collaborate in a session. They can view output and send input.
```
```suggestion
- **Peers** can join and collaborate in a session. They can view output and send input.
```

ptgott commented on docs/pages/setup/reference/moderated-sessions.mdx:
```diff
@@ -0,0 +1,115 @@
+
+A participant joining a session will always have one of three modes:
+- Peers can join and collaborate in a session. They can view output and send input.
+- Moderators can join and watch a session. They can view output and forcefully terminate the session at will.
```
```suggestion
- **Moderators** can join and watch a session. They can view output and forcefully terminate the session at will.
```

ptgott commented on docs/pages/setup/reference/moderated-sessions.mdx:
```diff
@@ -0,0 +1,115 @@
+A participant joining a session will always have one of three modes:
+- Peers can join and collaborate in a session. They can view output and send input.
+- Moderators can join and watch a session. They can view output and forcefully terminate the session at will.
+- Observers can join and watch a session. They cannot control the session in any way.
```
```suggestion
- **Observers** can join and watch a session. They cannot control the session in any way.
```

ptgott commented on docs/pages/setup/reference/moderated-sessions.mdx:
```diff
@@ -0,0 +1,115 @@
+- Moderators can join and watch a session. They can view output and forcefully terminate the session at will.
+- Observers can join and watch a session. They cannot control the session in any way.
+
+When joining a session with `tsh join` or `tsh kube join`, a mode can be specified with `--mode <mode>`
```
```suggestion
When joining a session with `tsh join` or `tsh kube join`, a user can specify a mode with the `--mode <mode>` flag
```
I think this sounds more direct in the active voice.

ptgott commented on docs/pages/setup/reference/moderated-sessions.mdx:
```diff
@@ -0,0 +1,115 @@
+- Observers can join and watch a session. They cannot control the session in any way.
+
+When joining a session with `tsh join` or `tsh kube join`, a mode can be specified with `--mode <mode>`
+where mode is one of `peer`, `moderator` or `observer`. By default, the mode is `peer` for SSH and
```
```suggestion
, where the mode is one of `peer`, `moderator` or `observer`. By default, the mode is `peer` for SSH and
```

ptgott commented on docs/pages/setup/reference/moderated-sessions.mdx:
```diff
@@ -0,0 +1,115 @@
+If no MFA input is received within 60 seconds, the user is kicked
+from the session which may pause it, if RBAC policies are no longer met.
+
+### Participant Modes
```
Should we make this an H2-level section since it doesn't have to do with configuring RBAC resources?

xacrimon replied:
I've liberally decided to convert most of the headers to H2, maybe It'd make sense to group Session kinds, Filters, example policies and MFA-based presence under a H2 named RBAC configuration and make them H3, what do you think?

xacrimon replied:
I've reworked headings now, take a look

xacrimon committed:
swap headers

xacrimon committed:
Apply suggestions from code review

Co-authored-by: Paul Gottschling <paul.gottschling@goteleport.com>

xacrimon committed:
mfa setting name

xacrimon committed:
link mfa

xacrimon commented on docs/pages/setup/reference/moderated-sessions.mdx:
```diff
@@ -0,0 +1,115 @@
+
+# Moderated Sessions
+
+Moderated Sessions makes use of RBAC to control access to and within sessions.
```
👍

xacrimon committed:
rewrite

ptgott commented on docs/pages/setup/reference/moderated-sessions.mdx:
```diff
@@ -0,0 +1,115 @@
+Allow policies are used to define what sessions a user can join
+and under what conditions they may join a session.
+
+## Configuring moderated sessions
```
My biggest issue with this section is that it is more like a guide than a reference. When reading a reference, I would expect a comprehensive set of available options (e.g., see our docs guide: https://goteleport.com/docs/docs/#reference-manuals). Instead, each section describes or illustrates a topic without being exhaustive.

I recommend reworking this section to have the following structure:
1. Add headers for `require_session_join` and `join_sessions`
1. Under each header, list the required/optional keys and their possible values (or the types of these values), as well as what these options configure. I suggest using a table.
1. Add a separate section explaining the filter syntax. The section should have the same heading level as the `require_session_join` and `join_sessions` sections. When we indicate the type of the `filter` option, we can link to that section.
1. Change "Session kinds" to use a table. We can link to this section when specifying the type of the `kinds` value.
1. Move the "Participant modes" section into the same heading level as the `join_sessions`/`require_session_join` sections. That way, when we indicate the type of the `modes` field, we can say that it's a list of participant modes, and link to that section.

As an illustration of what I mean, I think some parts of the Istio reference documentation do a good job at laying out tables of YAML config options and specifying the types of the values. In [this example](https://istio.io/latest/docs/reference/config/networking/gateway/#Gateway), we can see that the the `servers` field is a `Server[]`, and a separate section specifies the fields of `Server`.

ptgott reviewed (commented):
Other than the one comment I left, I think the guide looks good.

I think the reference needs to be reworked to be more comprehensive (I've left a more detailed comment below). If we want to get docs for this feature out sooner, I think we could move forward with the guide as a separate PR while the reference is being reworked. 

Let me know if you'd like me to work on a commit that implements my suggestions for the reference—I should have some time early next week to do this.

ptgott committed:
Add a conceptual guide to Moderated Sessions (#10094)

The tutorial-style guide created in previous commits did not
provide enough context into how the feature works and which
configuration options are available. Since the instructions in
the guide are mostly example configurations anyway, I thought it
would make sense to combine the guide and the reference while
removing the tutorial instructions and adding more detailed
guidance on available configuration options.

This guide now lives in docs/pages/access-controls. I thought
keeping it in docs/pages/setup/reference would not be appropriate,
since the references there are for broad approaches to interacting
with Teleport (configuration, dynamic resources, metrics, etc.),
rather than specific features.

We can plan to edit the dynamic resource reference to include
the config options mentioned in this guide.

xacrimon committed:
Merge branch 'master' into joel/moderated-sessions-preview-docs

xacrimon committed:
minor editorial changes

xacrimon committed:
mention count field

xacrimon committed:
update rbac reference

xacrimon committed:
cleanup + remove todo

xacrimon committed:
invites

xacrimon commented:
@ptgott Went ahead and filled out the conceptual guide you merged in. Please take another look when you have a moment.

xacrimon committed:
remove dangling link

xacrimon commented on docs/pages/access-controls/guides/moderated-sessions.mdx:
```diff
@@ -0,0 +1,173 @@
+
+The following are required options for `join_sessions`:
+
+|Option|Type|Description|
+|---|---|---|
+|`name`|String|The name of the allow policy|
+|`roles`|[]String|A list of names for Teleport roles that this policy applies to. Users with this role are eligible to join a Moderated Session under this policy.|
+|`kinds`|`[]`[Session kind](./moderated-sessions.mdx#session-kinds)|The kind of session that the policy applies to|
+|`modes`|`[]`[Participant mode](./moderated-sessions.mdx#participant-modes)|The participant mode that paplies to the user joining the Moderated Session under this policy|
```
A suggestion was made by @ptgott here to allow for filter expressions instead of a roles list if I read the comment in the conceptual guide PR correctly. I reasoned this in the original RFD and ended up sticking with the roles list to make the intended way to construct access hierarchies/systems using these policies more explicit. I feel like a filter expression gives flexibility that doesn't give us any better capabilities in return.

ptgott commented on docs/config.json:
```diff
@@ -88,7 +88,8 @@
             { "title": "Networking", "slug": "/setup/reference/networking/" },
-            { "title": "Enterprise License File", "slug": "/setup/reference/license/" }
+            { "title": "Enterprise License File", "slug": "/setup/reference/license/" },
+            { "title": "Moderated Sessions", "slug": "/setup/reference/moderated-sessions/" }
```
Let's delete this entry since the page was removed

ptgott commented on docs/pages/access-controls/reference.mdx:
```diff
@@ -130,6 +130,35 @@ spec:
+      # Defines the name of the policy. The name serves only as an
+      # identifier in logs and for organisation/categorisation.
+      - name: Auditor oversight
+        # Specifies an RBAC predicate that is used define
```
```suggestion
        # Specifies an RBAC predicate that is used to define
```

ptgott commented on docs/pages/access-controls/reference.mdx:
```diff
@@ -130,6 +130,35 @@ spec:
+      # Defines the name of the policy. The name serves only as an
+      # identifier in logs and for organisation/categorisation.
+      - name: Auditor oversight
+        # The list of roles of whom you can join their session.
```
Is there another way to phrase this?

ptgott commented on docs/pages/access-controls/guides/moderated-sessions.mdx:
```diff
@@ -0,0 +1,173 @@
+
+## Introduction
+
+Moderated sessions allows Teleport administrators to
```
```suggestion
Moderated Sessions allows Teleport administrators to
```

ptgott commented on docs/pages/access-controls/guides/moderated-sessions.mdx:
```diff
@@ -0,0 +1,173 @@
+### Use cases
+
+Moderated Sessions are useful in the following scenarios:
+- When you have stringent security requirements and need to have people watching over user-initiated sessions to a set of servers.
```
```suggestion
- When you have stringent security requirements and need to have people watching over user-initiated sessions on a set of servers.
```

ptgott commented on docs/pages/access-controls/guides/moderated-sessions.mdx:
```diff
@@ -0,0 +1,173 @@
+|`filter`|[Filter](./moderated-sessions.mdx#filters)|An expression that, if it evaluates to true for a given user, enables the user to be present in a Moderated Session|
+|`kinds`|`[]`[Session kind](./moderated-sessions.mdx/#session-kinds)|The kind of session that the policy applies to|
+|`modes`|`[]`[Participant mode](./moderated-sessions.mdx#participant-modes)|The participant mode that applies to the user joining the Moderated Session under this policy|
+|`count`|[Participant mode](./moderated-sessions.mdx#require-policy-count)|The amount of users that need to match the filter expression to satisfy the policy|
```
```suggestion
|`count`|Integer|The number of users that need to match the filter expression to satisfy the policy|
```

ptgott commented on docs/pages/access-controls/guides/moderated-sessions.mdx:
```diff
@@ -0,0 +1,173 @@
+
+#### Example
+
+The policy below specifies that a given role
```
```suggestion
The policy below specifies that the `prod-access` role
```

ptgott commented on docs/pages/access-controls/guides/moderated-sessions.mdx:
```diff
@@ -0,0 +1,173 @@
+## Configuring Moderated Sessions
+
+### `require_session_join`
+
```
```suggestion

`require_session_join` configures the require policies for a role. The value of 
`require_session_join` is a list of objects with the following fields.
```
Reading this now, I think we need some kind of introduction for this section.

ptgott commented on docs/pages/access-controls/guides/moderated-sessions.mdx:
````diff
@@ -0,0 +1,173 @@
+```
+
+### `join_sessions`
+
````
```suggestion

`join_sessions` configures the allow policies for a role. The value of 
`join_sessions` is a list of objects with the following fields.
```
Reading this now, I think we need some kind of introduction for this section.

ptgott commented:
@xacrimon This is a partial review—I need to resume reviewing this at the top of the "Filters" section in the guide—but I'm adding the comments I've made so far.

ptgott commented on docs/pages/access-controls/guides/moderated-sessions.mdx:
```diff
@@ -0,0 +1,173 @@
+
+### Filters
+
+A filter expression or predicate is a string statement used to define logic based on a set of input variables.
```
```suggestion
A filter expression is a string statement used to define logic based on a set of input variables.
```
We only say "filter expression" later on, so I thought we could remove "predicate" to simplify things.

ptgott commented on docs/pages/access-controls/guides/moderated-sessions.mdx:
```diff
@@ -0,0 +1,173 @@
+`user.name == "adam" || contains(user.roles, "cs-observe")` is a simple example of a filter expression
+that evaluates to true if the user is Adam or if the user has the trait `cs-observe`.
+
+Require policies can specify which users they consider as valid with a filter expression. They work like
```
```suggestion
Require policies can specify which users they consider as valid with a filter expression.
```
Suggested removing this sentence since we don't have good docs elsewhere on filter expressions.

ptgott commented on docs/pages/access-controls/guides/moderated-sessions.mdx:
```diff
@@ -0,0 +1,173 @@
+that evaluates to true if the user is Adam or if the user has the trait `cs-observe`.
+
+Require policies can specify which users they consider as valid with a filter expression. They work like
+other filters used within Teleport such with access requests. The filter context has a `user` object defined
```
```suggestion
 The filter context has a `user` object defined
```

Suggested removing this sentence since we don't have good docs elsewhere on filter expressions.

ptgott commented on docs/pages/access-controls/guides/moderated-sessions.mdx:
```diff
@@ -0,0 +1,173 @@
+Require and allow policies have to specify which sessions they apply to. Valid options are `ssh` and `k8s`.
+
+- `ssh` policies apply to all SSH sessions towards a node running the Teleport SSH server.
+- `k8s` policies apply to all kubernetes sessions to clusters connected to teleport.
```
```suggestion
- `k8s` policies apply to all Kubernetes sessions on clusters connected to Teleport.
```

ptgott commented on docs/pages/access-controls/guides/moderated-sessions.mdx:
```diff
@@ -0,0 +1,173 @@
+## Session invites
+
+Moderatated Sessions tracks sessions via `SessionTracker` resource. When starting
+an interactive SSH or Kubernetes session using `tsh ssh` or `tsh kube exec` respectively,
```
I wasn't sure how the mention of the `SessionTracker` resource relates to the CLI guidance in the next sentence.

ptgott commented on docs/pages/access-controls/guides/moderated-sessions.mdx:
````diff
@@ -0,0 +1,173 @@
+        modes: ['moderator', 'observer']
+```
+
+### Filters
````
Should we introduce all of the functions/operators in filter expressions here?

xacrimon replied:
Added

xacrimon committed:
Apply suggestions from code review

Co-authored-by: Paul Gottschling <paul.gottschling@goteleport.com>

xacrimon committed:
remove config entry

xacrimon committed:
Update docs/pages/access-controls/guides/moderated-sessions.mdx

Co-authored-by: Paul Gottschling <paul.gottschling@goteleport.com>

xacrimon committed:
feedback

xacrimon committed:
update v4 examples to v5

xacrimon committed:
fix accidental replace

ptgott commented on docs/pages/access-controls/guides/moderated-sessions.mdx:
```diff
@@ -0,0 +1,179 @@
+`equals(user.name, "adam") || contains(user.roles, "cs-observe")` is a simple example of a filter expression
+that evaluates to true if the user is Adam or if the user has the trait `cs-observe`.
+
+Require policies can specify which users they consider as valid with a filter expression.
```
Should we move this and the next sentence to the relevant table row in the section on configuring require policies?

ptgott commented on docs/pages/access-controls/guides/moderated-sessions.mdx:
```diff
@@ -0,0 +1,179 @@
+
+Require policies can specify which users they consider as valid with a filter expression.
+ The filter context has a `user` object defined
+which the set fields `roles` and `name`.
```
```suggestion
with the set fields `roles` and `name`.
```

xacrimon committed:
Update docs/pages/access-controls/guides/moderated-sessions.mdx

Co-authored-by: Paul Gottschling <paul.gottschling@goteleport.com>

xacrimon committed:
feedback

ptgott commented on docs/pages/access-controls/guides/moderated-sessions.mdx:
```diff
@@ -0,0 +1,178 @@
+|`name`|String|The name of the allow policy|
+|`roles`|[]String|A list of names for Teleport roles that this policy applies to. Users with this role are eligible to join a Moderated Session under this policy.|
+|`kinds`|`[]`[Session kind](./moderated-sessions.mdx#session-kinds)|The kind of session that the policy applies to|
+|`modes`|`[]`[Participant mode](./moderated-sessions.mdx#participant-modes)|The participant mode that paplies to the user joining the Moderated Session under this policy|
```
```suggestion
|`modes`|`[]`[Participant mode](./moderated-sessions.mdx#participant-modes)|The participant mode that applies to the user joining the Moderated Session under this policy|
```

ptgott commented on docs/pages/access-controls/guides/moderated-sessions.mdx:
```diff
@@ -0,0 +1,178 @@
+|`modes`|`[]`[Participant mode](./moderated-sessions.mdx#participant-modes)|The participant mode that applies to the user joining the Moderated Session under this policy|
+|`count`|Integer|The number of users that need to match the filter expression to satisfy the policy|
+
+Filter expressions allow for more detailed control over the scope of an allow policy or require policy.
```
On second thought, this and the next paragraph make more sense where they originally were, in the "Filters" section. This paragraph is a good example of using filters—which we need in the "Filters" section—and the paragraph beginning "Require policies can specify" only makes sense if you understand the filter syntax, i.e., at the end of the "Filters" section.

However, when we move these paragraphs  back to the "Filters" section, maybe we could use this order of sentences, since the example filter illustrates the `user.roles` and `user.name` fields (we'll need to unescape the code fences).

```markdown
Filter expressions allow for more detailed control over the scope of an allow policy or require policy.

Require policies can specify which users they consider as valid with a filter expression.
The filter context has a `user` object defined with the set fields `roles` and `name`.

Here is an example of a filter expression that evaluates to true if the user is Adam or if the user has the trait `cs-observe`:

\`\`\`
equals(user.name, "adam") || contains(user.roles, "cs-observe")
\`\`\`

```

xacrimon committed:
Update docs/pages/access-controls/guides/moderated-sessions.mdx

Co-authored-by: Paul Gottschling <paul.gottschling@goteleport.com>

xacrimon committed:
feedback

xacrimon committed:
fix links

xacrimon committed:
Merge branch 'master' into joel/moderated-sessions-preview-docs

xacrimon committed:
Merge branch 'master' into joel/moderated-sessions-preview-docs
